src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp

```diff
@@ -479,6 +566,185 @@ std::unique_ptr<MergeTreeBlockSizePredictor> MergeTreeBaseSelectProcessor::getSi
         data_part, Names(complete_column_names.begin(), complete_column_names.end()), sample_block);
 }
 
+
+MergeTreeBaseSelectProcessor::Status MergeTreeBaseSelectProcessor::performRequestToCoordinator(MarkRanges requested_ranges)
+{
+    String partition_id = task->data_part->info.partition_id;
+    String part_name;
+    String projection_name;
+
+    if (task->data_part->isProjectionPart())
+    {
+        part_name = task->data_part->getParentPart()->name;
+        projection_name  = task->data_part->name;
+    }
+    else
+    {
+        part_name = task->data_part->name;
+        projection_name = "";
+    }
+
+    PartBlockRange block_range
+    {
+        .begin = task->data_part->info.min_block,
+        .end = task->data_part->info.max_block
+    };
+
+    PartitionReadRequest request
+    {
+        .partition_id = std::move(partition_id),
+        .part_name = std::move(part_name),
+        .projection_name = std::move(projection_name),
+        .block_range = std::move(block_range),
+        .mark_ranges = std::move(requested_ranges)
+    };
+
+    AtomicStopwatch stop;
+
+    auto response = extension.value().callback(std::move(request));
+
+    LOG_TRACE(log, "Elapsed time to perform request: {}ns", stop.elapsed());
+
+    auto dump_mark_ranges = [&] (MarkRanges ranges)
+    {
+        String result;
+        for (const auto & range: ranges)
+            result += fmt::format("[{} {}], ", range.begin, range.end);
+        result += '\n';
+        return result;
+    };
+
+    LOG_TRACE(log, (response.denied ? "denied" : "accepted"), dump_mark_ranges(response.mark_ranges));
+
+    task->mark_ranges = std::move(response.mark_ranges);
+
+    if (response.denied)
+        return Status::Denied;
+
+    if (task->mark_ranges.empty())
+        return Status::Denied;
+
+    assert(!task->mark_ranges.empty());
+
+    finalizeNewTask();
+
+    return Status::Accepted;
+}
+
+
+template <typename Predicate>
+void MergeTreeBaseSelectProcessor::fillBufferedRanged(MergeTreeReadTask * current_task, Predicate && predicate)
+{
+    size_t sum_average_marks_size = 0;
+    /// getColumnSize is not fully implemented for compact parts
+    if (task->data_part->getType() == IMergeTreeDataPart::Type::COMPACT)
+    {
+        sum_average_marks_size = 8UL * 1024 * 1024 * 10; // 10 MiB
```

Comment:
Please, declare this constant separately and add a comment why it is 10MiB. You used it 3 times.